File: Backend/Backend/Models/Producto.cs
namespace Backend.Models;

/// <summary>
/// Producto del catalogo.
///
/// El stock de un producto se lleva SIEMPRE en su unidad base: kilos para el
/// azucar, unidades (botellas) para el aceite. Comprar un saco de 50 suma 50
/// kilos y vender 3 kilos resta 3, sin importar en que presentacion se hizo la
/// operacion. Las equivalencias viven en <see cref="ProductoPresentacion"/>.
///
/// Cada medida de un envasado es un producto distinto: "Aceite Primor 1 L" y
/// "Aceite Primor 900 ml" son dos productos, porque tienen stock y precio
/// propios. Lo que comparten es marca y categoria.
/// </summary>
public class Producto
{
    public int Id { get; set; }

    /// <summary>Codigo interno con el que se busca el producto. No se repite.</summary>
    public string Codigo { get; set; } = string.Empty;

    public string Nombre { get; set; } = string.Empty;
    public string? Descripcion { get; set; }

    public int? CategoriaId { get; set; }
    public Categoria? Categoria { get; set; }

    public int? MarcaId { get; set; }
    public Marca? Marca { get; set; }

    /// <summary>Unidad en la que se guarda el stock: KG, UND, LT.</summary>
    public int UnidadBaseId { get; set; }
    public UnidadMedida? UnidadBase { get; set; }

    /// <summary>
    /// Cuanto contiene el envase, para productos envasados: 900 junto a la
    /// unidad ML. Es informativo — sirve para comparar precio por litro — y no
    /// interviene en el stock.
    /// </summary>
    public decimal? Contenido { get; set; }

    public int? ContenidoUnidadId { get; set; }
    public UnidadMedida? ContenidoUnidad { get; set; }

    /// <summary>
    /// Lo que SUELE costar una unidad base, para tenerlo a mano al cotizar y
    /// al recibir mercaderia.
    ///
    /// Es una referencia tuya, NO el costo del inventario: ese lo fija cada
    /// entrada, porque el proveedor sube y baja el precio y pueden convivir
    /// varios costos del mismo producto. Sirve como valor sugerido cuando una
    /// entrada pide costo, y para ver el margen de una lista de precios sin
    /// tener stock todavia.
    /// </summary>
    public decimal? CostoReferencia { get; set; }

    /// <summary>
    /// A cuánto SUELE venderse una unidad base.
    ///
    /// Es el precio de respaldo: cuando el pedido no lleva lista de precios, o la lista elegida no
    /// tiene cargada esa presentación, es este el que sale. Así ningún producto se vende en cero
    /// por no haber terminado de armar las listas.
    ///
    /// La lista MANDA sobre él: si la lista tiene precio para esa presentación y esa cantidad, se
    /// cobra el de la lista. Este es el piso común y la referencia al armar una lista nueva.
    /// </summary>
    public decimal? PrecioReferencia { get; set; }

    /// <summary>
    /// Cuánto pesa UNA unidad base, en kilos.
    ///
    /// De aquí sale el peso de cualquier cantidad sin tener que anotarlo en cada presentación: una
    /// caja de 12 botellas de 0.92 kg pesa 11.04, y un pedido de 10 cajas, 110.4. Sirve para saber
    /// qué carga lleva el camión. Vacío en lo que no se pesa.
    /// </summary>
    public decimal? PesoUnidadBase { get; set; }

    /// <summary>
    /// Si el producto paga IGV. La mayoría sí — los exonerados (varios
    /// alimentos de primera necesidad) son la excepción.
    ///
    /// El <see cref="PrecioReferencia"/> y los precios de presentación YA
    /// incluyen el IGV cuando el producto es afecto: se cobra tal cual se
    /// escribió, sin sumarle nada encima. Este campo no cambia cómo se cobra
    /// — es para separar el IGV al momento de reportar o emitir un
    /// comprobante.
    /// </summary>
    public bool AfectoIgv { get; set; }

    /// <summary>Servicios y similares no descuentan stock.</summary>
    public bool ControlaStock { get; set; } = true;

    /// <summary>Aviso de reposicion, en unidad base.</summary>
    public decimal StockMinimo { get; set; }

    public bool Activo { get; set; } = true;
    public DateTime FechaCreacion { get; set; } = DateTime.UtcNow;

    public List<ProductoPresentacion> Presentaciones { get; set; } = [];
}

File: titanic/lib/features/maestros/datos/producto.dart
/// Una forma de comprar o vender un producto: saco de 50, caja x12.
class Presentacion {
  const Presentacion({
    required this.id,
    required this.unidadId,
    required this.unidad,
    required this.nombre,
    required this.factor,
    required this.esBase,
    required this.esCompra,
    required this.esVenta,
    this.precioPorPresentacion = false,
    required this.activo,
  });

  final int id;
  final int unidadId;
  final String unidad;
  final String nombre;

  /// Cuantas unidades base equivale. Un saco de 50 kg tiene factor 50.
  final double factor;

  /// La de factor 1: la crea sola el backend y no se elimina ni cambia.
  final bool esBase;
  final bool esCompra;
  final bool esVenta;

  /// Los PDF de ajustes, transferencias y prestamos sacan la linea en esta
  /// presentacion ("150 cajas × S/ 83.00") y no en unidad base ("1,800 × S/ 6.92").
  /// Lo arma el servidor: la app solo lo lee y lo devuelve. Apagado por defecto
  /// —y en la base no se ofrece, por unidad base ya ES por presentacion—.
  final bool precioPorPresentacion;
  final bool activo;

  factory Presentacion.desdeJson(Map<String, dynamic> json) => Presentacion(
    id: json['id'] as int,
    unidadId: json['unidadId'] as int,
    unidad: json['unidad'] as String? ?? '',
    nombre: json['nombre'] as String? ?? '',
    factor: (json['factor'] as num?)?.toDouble() ?? 1,
    esBase: json['esBase'] as bool? ?? false,
    esCompra: json['esCompra'] as bool? ?? true,
    esVenta: json['esVenta'] as bool? ?? true,
    precioPorPresentacion: json['precioPorPresentacion'] as bool? ?? false,
    activo: json['activo'] as bool? ?? true,
  );

  /// Cuerpo de POST/PUT de una presentacion.
  ///
  /// El endpoint REEMPLAZA la presentacion con lo que llega y un campo ausente
  /// se guarda como apagado: por eso el marcador viaja siempre, con el valor
  /// que ya tenia, para no borrar el que se puso desde la web.
  Map<String, dynamic> aJson() => {
    'unidadId': unidadId,
    'nombre': nombre,
    'factor': factor,
    'esCompra': esCompra,
    'esVenta': esVenta,
    'precioPorPresentacion': precioPorPresentacion,
    'activo': activo,
  };
}

/// Producto: lo que se compra y se vende.
class Producto {
  const Producto({
    required this.id,
    required this.codigo,
    required this.nombre,
    this.descripcion,
    this.categoriaId,
    this.categoria,
    this.marcaId,
    this.marca,
    required this.unidadBaseId,
    required this.unidadBase,
    this.costoReferencia,
    this.precioReferencia,
    this.afectoIgv = false,
    required this.controlaStock,
    required this.stockMinimo,
    this.pesoUnidadBase,
    required this.activo,
    required this.presentaciones,
    this.tieneMovimientos = false,
  });

  final int id;
  final String codigo;
  final String nombre;
  final String? descripcion;

  /// Ya tiene stock, kardex o costos: cambiar su unidad base no los convierte.
  final bool tieneMovimientos;
  final int? categoriaId;
  final String? categoria;
  final int? marcaId;
  final String? marca;
  final int unidadBaseId;

  /// Codigo de la unidad en la que se lleva el stock: KG, UND, LT.
  final String unidadBase;

  /// Lo que suele costar una unidad base. Referencia, no el costo del stock.
  final double? costoReferencia;

  /// A cuánto suele venderse una unidad base: el respaldo cuando la lista no tiene precio.
  /// Ya incluye el IGV si el producto es afecto.
  final double? precioReferencia;

  /// Si paga IGV. Los precios ya lo incluyen: no se le suma nada encima.
  final bool afectoIgv;

  final bool controlaStock;
  final double stockMinimo;

  /// Kilos que pesa UNA unidad base. Null si el producto no se pesa.
  ///
  /// De aqui sale el peso de cualquier cantidad sin anotarlo en cada
  /// presentacion: si la botella pesa 0.92, la caja de 12 pesa 11.04 y diez
  /// cajas 110.4.
  final double? pesoUnidadBase;

  final bool activo;
  final List<Presentacion> presentaciones;

  /// Texto contra el que se busca en la lista.
  String get buscable =>
      '$codigo $nombre ${categoria ?? ''} ${marca ?? ''}'.toLowerCase();

  factory Producto.desdeJson(Map<String, dynamic> json) => Producto(
    id: json['id'] as int,
    tieneMovimientos: json['tieneMovimientos'] as bool? ?? false,
    codigo: json['codigo'] as String? ?? '',
    nombre: json['nombre'] as String? ?? '',
    descripcion: json['descripcion'] as String?,
    categoriaId: json['categoriaId'] as int?,
    categoria: json['categoria'] as String?,
    marcaId: json['marcaId'] as int?,
    marca: json['marca'] as String?,
    unidadBaseId: json['unidadBaseId'] as int,
    unidadBase: json['unidadBase'] as String? ?? '',
    costoReferencia: (json['costoReferencia'] as num?)?.toDouble(),
    precioReferencia: (json['precioReferencia'] as num?)?.toDouble(),
    afectoIgv: json['afectoIgv'] as bool? ?? false,
    controlaStock: json['controlaStock'] as bool? ?? true,
    stockMinimo: (json['stockMinimo'] as num?)?.toDouble() ?? 0,
    pesoUnidadBase: (json['pesoUnidadBase'] as num?)?.toDouble(),
    activo: json['activo'] as bool? ?? true,
    presentaciones: (json['presentaciones'] as List? ?? const [])
        .map((e) => Presentacion.desdeJson(e as Map<String, dynamic>))
        .toList(),
  );
}
